### PR TITLE
Fixed home button double-press issue, Fixed home profile latching actions issue

### PR DESCRIPTION
**Pull request by marcos-diaz**
#126 and #136

### CODE REVIEW

File: src/profile.c [hunk 1]
```diff
@@ -259,13 +260,19 @@ void profile_update_leds() {
     }
 }
 
-void profile_report_active() {
-    if (profile_pending_reboot && !home_is_active) config_reboot();
+// Check if profile has been requested to be reset (usually after profile change).
```
Comment (marcos-diaz) [inline src/profile.c]: Git does a poor job here, this is a new function extracted away from the function below

File: src/hid.c [hunk 2]
```diff
@@ -20,7 +20,7 @@ bool synced_gamepad = false;
 uint16_t alarms = 0;
 alarm_pool_t *alarm_pool;
 
-uint8_t state_matrix[256] = {0,};
+int8_t state_matrix[256] = {0,};
```
Comment (przymi) [inline src/hid.c]: Add comment that negative means an error condition

File: src/headers/hid.h [hunk 3]
```diff
@@ -263,7 +263,7 @@
 
 
 void hid_thanks();
-void hid_matrix_reset();
+void hid_matrix_reset(uint8_t exclude);
```
Comment (przymi) [inline src/headers/hid.h]: Where do I find some documentation on this hid_matrix?

File: src/hid.c [hunk 4]
```diff
@@ -1,6 +1,33 @@
...
+                    --> Profile -->             --> XInput gamepad
+                    --> Profile -->             --> Internal procedures
+
+At the end of each cycle (determined by the polling rate) the HID layer checks
+if the potential new report is different from the last report sent to the
+interfaces (USB keyboard, USB mouse, gamepad...), and sends the report if
+required.
+
+The state matrix is a representation of all the actions that could be sent
+(output) and internal operations (procedures). It keep track of how many
```
Comment (przymi) [inline src/hid.c]: Actions are coming from the user. What are internal operations in this context?

File: src/hid.c [hunk 5]
```diff
@@ -1,6 +1,33 @@
...
+references to these actions are active. As a simplification: Button presses
```
Comment (przymi) [inline src/hid.c]: How many references (to prior actions?) are kept and where?

File: src/hid.c [hunk 6]
```diff
@@ -1,6 +1,33 @@
...
+                    --> Profile -->             --> HID mouse
+                    --> Profile -->             --> XInput gamepad
+                    --> Profile -->             --> Internal procedures
+
+At the end of each cycle (determined by the polling rate) the HID layer checks
+if the potential new report is different from the last report sent to the
+interfaces (USB keyboard, USB mouse, gamepad...), and sends the report if
+required.
+
+The state matrix is a representation of all the actions that could be sent
```
Comment (przymi) [inline src/hid.c:20]: So this is the HID matrix.

File: src/hid.c [hunk 7]
```diff
@@ -30,9 +57,10 @@ double gamepad_ry = 0;
 double gamepad_lz = 0;
 double gamepad_rz = 0;
 
-void hid_matrix_reset() {
-    for(uint8_t i=0; i<255; i++) {
-        state_matrix[i] = 0;
+void hid_matrix_reset(uint8_t keep) {
```
Comment (przymi) [inline src/hid.c:61]: What is "keep"?
According to the description above, "state matrix is usually re-initialized (reset) to zeros when the user changes the active profile"

File: src/hid.c [hunk 8]
```diff
@@ -105,10 +133,12 @@ void hid_release(uint8_t key) {
     else if (key == MOUSE_SCROLL_DOWN) return;
     else if (key >= PROC_INDEX) hid_procedure_release(key);
     else {
-        state_matrix[key] -= 1;
-        if (key >= GAMEPAD_INDEX) synced_gamepad = false;
-        else if (key >= MOUSE_INDEX) synced_mouse = false;
-        else synced_keyboard = false;
+        if (state_matrix[key] > 0) {  // Do not allow to wrap / go negative.
```
Comment (przymi) [inline src/hid.c:137]: What is the difference between keys and actions in the index of state_matrix?
Would not all of these have been reset above?

File: src/hid.c [hunk 9]
```diff
@@ -105,10 +133,12 @@ void hid_release(uint8_t key) {
...
+            state_matrix[key] -= 1;
+            if (key >= GAMEPAD_INDEX) synced_gamepad = false;
+            else if (key >= MOUSE_INDEX) synced_mouse = false;
```
Comment (przymi) [inline src/hid.c:140]: These two are already false after line 66

File: src/profile.c [hunk 10]
```diff
@@ -259,13 +260,19 @@ void profile_update_leds() {
...
+void profile_check_reset() {
```
Comment (przymi) [inline src/profile.c]: THis is not checking if there is / or should be a reset, but it is executing a reset

File: src/profile.c [hunk 11]
```diff
@@ -259,13 +260,19 @@ void profile_update_leds() {
...
     if (pending_reset) {
-        hid_matrix_reset();
+        hid_matrix_reset(pending_reset_keep);
```
Comment (przymi) [inline src/profile.c]: ???

File: src/profile.c [hunk 12]
```diff
@@ -259,13 +260,19 @@ void profile_update_leds() {
...
         profile_reset_all();
         pending_reset = false;
+        pending_reset_keep = 0;
     }
+}
+
+void profile_report_active() {
```
Comment (przymi) [inline src/profile.c:273]: What is this supposed to do?

File: src/profile.c [hunk 13]
```diff
@@ -287,10 +294,12 @@ void profile_set_home_gamepad(bool state) {
     if (state) {
         led_static_mask(LED_NONE);
         led_set_mode(LED_MODE_ENGAGE);
+        home_is_active = false;  // Force normal home profile off.
```
Comment (przymi) [inline src/profile.c]: Is this the home profile? I.e. only active while the home button is pressed?

**Review by przymi — CHANGES_REQUESTED** [review]

File: src/hid.c (hunk 7 above)
Comment (marcos-diaz) [reply src/hid.c:61]: The high level documentation cannot explain all the implementation corner cases, I will add more code comments instead.

File: src/hid.c (hunk 8 above)
Comment (marcos-diaz) [reply src/hid.c:137]: That's legacy naming from when the only actions were keys (in the context of key presses and releases). 

Fixing the terminology would require renaming a lot of variables and function names across the whole project. Out of the scope of this PR.

File: src/hid.c (hunk 9 above)
Comment (marcos-diaz) [reply src/hid.c:140]: There are different functions with different triggers and usecases

File: src/profile.c (hunk 13 above)
Comment (marcos-diaz) [reply src/profile.c]: Actually the line was not needed, was removed

**Pull request merged**